File: app/src/main/java/pl/laskowski/marcin/solving/deduction/algorithm/HiddenValuesAlgorithm.java
package pl.laskowski.marcin.solving.deduction.algorithm;

import pl.laskowski.marcin.model.Region;
import pl.laskowski.marcin.solving.deduction.combinations.CollectionCombinator;
import pl.laskowski.marcin.model.Field;
import pl.laskowski.marcin.model.SudokuHintGrid;

import java.util.*;

/**
 * Created by Marcin Laskowski.
 */

public class HiddenValuesAlgorithm extends DeductionAlgorithm {

    public static class Factory implements DeductionAlgorithm.Factory {

        private final int size;

        public Factory(int size) {
            this.size = size;
        }

        public HiddenValuesAlgorithm instance(Set<Region> regions, SudokuHintGrid sudokuHintGrid) {
            return new HiddenValuesAlgorithm(regions, sudokuHintGrid, size);
        }

    }

    private final CollectionCombinator combinator;
    private final int size;

    private HiddenValuesAlgorithm(Set<Region> regions, SudokuHintGrid possibilities, int size) {
        super(regions, possibilities);
        this.combinator = new CollectionCombinator(size);
        this.size = size;
    }

    @Override
    protected void solve(Region region) {
        Set<Integer> allHintValues = possibilities.forRegion(region);
        combinator.iterate(allHintValues, result -> {
            Set<Integer> values = new HashSet<>(result);
            int fields = countFieldsThatContainsValues(region, values);
            if (fields == size) {
                removeAllOtherValues(region, values);
            }
        });
    }

    private int countFieldsThatContainsValues(Region region, Set<Integer> values) {
        int sum = 0;
        for (Field field : region) {
            Set<Integer> hints = possibilities.forField(field);
            if (!Collections.disjoint(hints, values)) sum++;
        }
        return sum;
    }

    private void removeAllOtherValues(Region region, Set<Integer> values) {
        for (Field field : region) {
            Set<Integer> hints = possibilities.forField(field);
            int initialSize = hints.size();
            if (!Collections.disjoint(hints, values)) {
                hints.removeIf(element -> !values.contains(element));
            }
            if (initialSize != hints.size()) {
                possibilities.replace(field, hints);
                notifyChange();
            }
        }
    }

}

File: src/main/kotlin/io/github/ilikeyourhat/kudoku/solving/deduction/algorithm/HiddenValuesAlgorithm.kt
package io.github.ilikeyourhat.kudoku.solving.deduction.algorithm

import io.github.ilikeyourhat.kudoku.model.Region
import io.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid
import io.github.ilikeyourhat.kudoku.solving.deduction.combinations.CollectionCombinator
import java.util.Collections

class HiddenValuesAlgorithm(
    regions: List<Region>,
    possibilities: SudokuHintGrid,
    private val size: Int
) : DeductionAlgorithm(regions, possibilities) {

    class Factory(private val size: Int) : DeductionAlgorithm.Factory {
        override fun instance(regions: List<Region>, possibilities: SudokuHintGrid): HiddenValuesAlgorithm {
            return HiddenValuesAlgorithm(regions, possibilities, size)
        }
    }

    private val combinator = CollectionCombinator(size)

    override fun solve(region: Region): Boolean {
        var changed = false
        val allHintValues = possibilities.forRegion(region)
        combinator.iterate(allHintValues) { result ->
            val values = result.toSet()
            val fields = countFieldsThatContainsValues(region, values)
            if (fields == size) {
                changed = changed or removeAllOtherValues(region, values)
            }
        }
        return changed
    }

    private fun countFieldsThatContainsValues(region: Region, values: Set<Int>): Int {
        var sum = 0
        for (field in region) {
            val hints = possibilities.forField(field)
            if (!Collections.disjoint(hints, values)) {
                sum++
            }
        }
        return sum
    }

    private fun removeAllOtherValues(region: Region, values: Set<Int>): Boolean {
        var changed = false
        for (field in region) {
            val hints = possibilities.forField(field).toMutableSet()
            val initialSize = hints.size
            if (!Collections.disjoint(hints, values)) {
                hints.removeIf { !values.contains(it) }
            }
            if (initialSize != hints.size) {
                possibilities.replace(field, hints)
                changed = true
            }
        }
        return changed
    }
}
